src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1385,6 +1388,156 @@ static LogicalResult getVariadicSimdCodeFullyFlattened(
+  OpFusionHelper(
+      mlir::ConversionPatternRewriter &rewriter, mlir::Operation *startOp)
+      : op_(startOp), rewriter_(rewriter), fusibleOps_(), fuseEmitFunctions_() {
+  }
+
+  bool checkFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+      SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+
+    // Notice: Though ClipOp is classified as unary element op in this file,
+    // ClipOp requires one required input and two optional input
+
+    return enqueueFusedOp<mlir::ONNXAbsOp, mlir::ONNXAtanOp, mlir::ONNXCastOp,
+        mlir::ONNXCeilOp, mlir::ONNXCosOp, mlir::ONNXCoshOp,
+        mlir::ONNXDequantizeLinearOp, mlir::ONNXEluOp, mlir::ONNXErfOp,
+        mlir::ONNXAcosOp, mlir::ONNXAcoshOp, mlir::ONNXAsinOp,
+        mlir::ONNXAsinhOp, mlir::ONNXAtanhOp, mlir::ONNXExpOp,
+        mlir::ONNXFloorOp, mlir::ONNXHardSigmoidOp, mlir::ONNXIsInfOp,
+        mlir::ONNXIsNaNOp, mlir::ONNXLeakyReluOp, mlir::ONNXLogOp,
+        mlir::ONNXNegOp, mlir::ONNXNotOp, mlir::ONNXReciprocalOp,
+        mlir::ONNXReluOp, mlir::ONNXRoundOp, mlir::ONNXSeluOp,
+        mlir::ONNXSigmoidOp, mlir::ONNXSignOp, mlir::ONNXSinOp,
+        mlir::ONNXSinhOp, mlir::ONNXSoftplusOp, mlir::ONNXSoftsignOp,
+        mlir::ONNXSqrtOp, mlir::ONNXTanOp, mlir::ONNXTanhOp>(
+        op, fusibleOps, fuseEmitFunctions);
+  }
+
+  bool fusibleOpsIsEmpty() { return fusibleOps_.size() == 0; }
+
+  // This function starts for elementwise operation, op.
+  // A successor op (user) is fusible if it's the only user and an unary
+  // elementwise Op. The Op and its EmitScalarOpFor<T> are recorded into
+  // the vector.
+  void findFusibleOps() {
+    Operation *currentProducer = op_;
+    while (currentProducer->hasOneUse()) {
+      // Check the users is an unary elementwise op
+      // The right solution, I think, is to define EmitScalarOpFor as
+      // an interface of unary elementwise Ops.
+      // In the draft PR, variadic template is used to iterate through
+      // the possible ONNX Ops.
+      Operation *user = *currentProducer->getUsers().begin();
+      if (!checkFusedOp(user, fusibleOps_, fuseEmitFunctions_))
+        break;
+
+      currentProducer = user;
+    }
+
+    if (!fusibleOpsIsEmpty())
+      LLVM_DEBUG({
+        llvm::dbgs() << "unary op fused: " << fusibleOps_.size() << "\n";
+      });
+  }
+
+  // After fusion, the only store is for the last Op.
+  // Therefore, the allocation should be the output of the last Op
+  MemRefType updateOutputType(MemRefType outputType) {
+    if (!fusibleOpsIsEmpty()) {
+      Operation *lastOp = fusibleOps_[fusibleOps_.size() - 1];
+      return MemRefType::get(
+          outputType.getShape(), lastOp->getResults()[0]
+                                     .getType()
+                                     .template cast<ShapedType>()
+                                     .getElementType());
+    } else {
+      return outputType;
+    }
+  }
+
+  // Emit fusion Ops
+  Value emitFuseOps(Value finalResult) {
```

Comment:
`emitFuseOps` may not fuse any ops, depending whether `findFusibleOps`. Suggest to use `tryEmitFuseOp` for the intention that it may succeed or fail.

Reply:
All the operation defined on OpFusionHelper works when the fusible Op list is empty. I would like to regard that empty list is one situation for OpFusionHelper. Alex also pointed out to add explicit if statement for empty fusible Op list. However, I prefer the uniform way to handle fusion list empty and non-empty. Using "try" may lead to the wonder of failure handling. I kept the name of this method and changed the name of the other two methods.